### PR TITLE
DPL: make sure that const services use the const API

### PR DESCRIPTION
**Pull request by ktf**

### CODE REVIEW

**Comment by matthiasrichter** [comment]
very good, I can still do the mistake to forget the const, e.g.
```
adaptStateful([inputSpecs](CallbackService& callbacks, RawDeviceService& rds, DeviceSpec& deviceSpec)
```

This throws a runtime error but there no reason printed. Although the runtime_error calls in `ServiceRegistry` have an error message, of course. So I don't know where this message got lost.

```
[7029:dpl-sink]: [17:43:10][STATE] Exiting FairMQ state machine
[7029:dpl-sink]: [17:43:10][ERROR] Unhandled o2::framework::runtime_error reached the top of main of stage/tests/o2-test-framework-ExternalFairMQDeviceWorkflow, device shutting down.
[7029:dpl-sink]:  Reason: 
[7029:dpl-sink]:  Backtrace follow: 
[7029:dpl-sink]: 
[7029:dpl-sink]: /home/o2/alisw/sw/BUILD/O2-latest/O2/stage/lib/libO2FrameworkFoundation.so(_ZN2o29framework15runtime_error_fEPKcz+0x146)[0x7f10f48b9395]
[7029:dpl-sink]: stage/tests/o2-test-framework-ExternalFairMQDeviceWorkflow(_ZNK2o29framework15ServiceRegistry3getEjmNS0_11ServiceKindEPKc+0x2a6)[0x46b800]
[7029:dpl-sink]: /home/o2/alisw/sw/BUILD/O2-latest/O2/stage/lib/libO2Framework.so(_ZNK2o29framework15ServiceRegistry3getINS0_10DeviceSpecEEERT_v+0x5e)[0x7f10f63cccd4]
[7029:dpl-sink]: /home/o2/alisw/sw/BUILD/O2-latest/O2/stage/lib/libO2Framework.so(_ZN2o29framework20ContextElementTraitsINS0_10DeviceSpecEE3getERNS0_11InitContextE+0x20)[0x7f10f63cc21a]
[7029:dpl-sink]: /home/o2/alisw/sw/BUILD/O2-latest/O2/stage/lib/libO2Framework.so(_ZZN2o29framework14adaptStatefulFIJNS0_15CallbackServiceENS0_16RawDeviceServiceENS0_10DeviceSpecEEEESt8functionIFS5_IFvRNS
0_17ProcessingContextEEERNS0_11InitContextEEES5_IFS9_DpRT_EEENKUlSB_E_clESB_+0x29)[0x7f10f63cc171]
[7029:dpl-sink]: /home/o2/alisw/sw/BUILD/O2-latest/O2/stage/lib/libO2Framework.so(_ZSt13__invoke_implISt8functionIFvRN2o29framework17ProcessingContextEEERZNS2_14adaptStatefulFIJNS2_15CallbackServiceENS2_1
6RawDeviceServiceENS2_10DeviceSpecEEEES0_IFS6_RNS2_11InitContextEEES0_IFS6_DpRT_EEEUlSC_E_JSC_EET_St14__invoke_otherOT0_DpOT1_+0x42)[0x7f10f63ceb72]
[
```

File: Framework/Core/include/Framework/ServiceRegistryHelpers.h [hunk 1]
```diff
@@ -40,6 +40,8 @@ struct ServiceRegistryHelpers {
     // This only works for concrete implementations of the type T.
     // We need type elision as we do not want to know all the services in
     // advance
+    static_assert(std::is_const_v<I> == false,
+                  "Service interface must be not const");
```
Comment (matthiasrichter) [inline Framework/Core/include/Framework/ServiceRegistryHelpers.h]:
```suggestion
                  "Service interface must not be const if service object is not const");
```

File: Framework/Core/include/Framework/ServiceRegistryHelpers.h [hunk 2]
```diff
@@ -53,9 +55,11 @@ struct ServiceRegistryHelpers {
     // This only works for concrete implementations of the type T.
     // We need type elision as we do not want to know all the services in
     // advance
+    static_assert(std::is_const_v<I> == true,
+                  "Service interface must be const");
```
Comment (matthiasrichter) [inline Framework/Core/include/Framework/ServiceRegistryHelpers.h]:
```suggestion
                  "Service interface must be const if service object is const");
```

**Review by matthiasrichter — DISMISSED** [review]
The error message can be a little more precise

**Comment by ktf** [comment]
Merging this. I will check why the message is not printed.

**Pull request merged**